pydantic/_internal/_core_metadata.py

```diff
@@ -1,84 +1,93 @@
 from __future__ import annotations as _annotations
 
-import typing
-from typing import Any, cast
+from typing import TYPE_CHECKING, Any, cast
 
-import typing_extensions
+from typing_extensions import TypedDict
 
-if typing.TYPE_CHECKING:
+if TYPE_CHECKING:
     from pydantic_core import CoreSchema
 
+    from ..config import JsonDict, JsonSchemaExtraCallable
     from ._schema_generation_shared import (
-        CoreSchemaOrField,
         GetJsonSchemaFunction,
     )
 
 
-class CoreMetadata(typing_extensions.TypedDict, total=False):
+class CoreMetadata(TypedDict, total=False):
     """A `TypedDict` for holding the metadata dict of the schema.
 
     Attributes:
-        pydantic_js_functions: List of JSON schema functions.
+        pydantic_js_functions: List of JSON schema functions that resolve refs during application.
+        pydantic_js_annotation_functions: List of JSON schema functions that don't resolve refs during application.
         pydantic_js_prefer_positional_arguments: Whether JSON schema generator will
             prefer positional over keyword arguments for an 'arguments' schema.
+        pydantic_js_input_core_schema: Schema associated with the input value for the associated
+            custom validation function. Only applies to before, plain, and wrap validators.
+        pydantic_js_udpates: key / value pair updates to apply to the JSON schema for a type.
+        pydantic_js_extra: WIP, either key/value pair updates to apply to the JSON schema, or a custom callable.
+
+    TODO: Perhaps we should move this structure to pydantic-core. At the moment, though,
+    it's easier to iterate on if we leave it in pydantic until we feel there is a semi-stable API.
+
+    TODO: It's unfortunate how functionally oriented JSON schema generation is, especially that which occurs during
+    the core schema generation process. It's inevitable that we need to store some json schema related information
+    on core schemas, given that we generate JSON schemas directly from core schemas. That being said, debugging related
+    issues is quite difficult when JSON schema information is disguised via dynamically defined functions.
     """
 
     pydantic_js_functions: list[GetJsonSchemaFunction]
     pydantic_js_annotation_functions: list[GetJsonSchemaFunction]
+    pydantic_js_prefer_positional_arguments: bool
+    pydantic_js_input_core_schema: CoreSchema
+    pydantic_js_updates: JsonDict
+    pydantic_js_extra: JsonDict | JsonSchemaExtraCallable
+
+
+def update_core_metadata(
+    core_metadata: Any,
+    /,
+    pydantic_js_functions: list[GetJsonSchemaFunction] | None = None,
+    pydantic_js_annotation_functions: list[GetJsonSchemaFunction] | None = None,
+    pydantic_js_prefer_positional_arguments: bool | None = None,
+    pydantic_js_input_core_schema: CoreSchema | None = None,
```

Comment:
This is probably overkill (these params specifically). Happy to remove. The others make more sense given that they need special update logic.